File: Wordplay/js/10.js
// What are all of the words that have all 5 vowels, in alphabetical order?

let main = () => {

    const fs = require("fs");
    let buffer_arry = fs.readFileSync("Wordplay/sowpods.txt", "utf-8");
    let in_file = buffer_arry.split("\n");
    let vowel_arry = ["A", "E", "I", "O", "U"];

    // Function to determine if word includes all 5 vowels
    let determine_all_vowels = (word) => {

        for(let i = 0; i < vowel_arry.length; i++) {            
            if(word.includes(vowel_arry[i]) == false) {
                return false;
            };
        };
        return true;
    };

    // Function to determine if all vowels are in alphabetical order
    let determine_order = (word) => {

        let all_vowels = [];
        let ordered_vowels = [];
        let word_len = word.length;

        // Push all vowels in word into an arry
        for(let i = 0; i < word_len; i++) {
            for(let j = 0; j < vowel_arry.length; j++) {
                if(word[i] == vowel_arry[j]) {
                    all_vowels.push(word[i])
                };
            };
        };

        // Remove duplicates from that array
        for(let i = 0; i < all_vowels.length; i++) {
            if(ordered_vowels.includes(all_vowels[i]) == false) {
                ordered_vowels.push(all_vowels[i]);
            };
        };

        // Check if vowels are in aphabetical order
        for(let i = 0; i < vowel_arry.length; i++) {
            if(ordered_vowels[i] != vowel_arry[i]) {
                return
            };
        };

        // If word meets all specifications print it out
        console.log(word);
    };


    // Main fucnction call
    for(let i = 0; i < in_file.length; i++) {
        if(determine_all_vowels(in_file[i])) {
            determine_order(in_file[i]);
        };
    };
};

main();

File: Wordplay/python/10.py
# What are all of the words that have all 5 vowels, in alphabetical order?

def main():
    in_file = open("../sowpods.txt", "r")

    def check_vowels(word):
        word_len = len(word) - 1 # -1 for escape character
        start_index = - 1
        vowel_list = []
        vowel_list_ol = []
        vowels = ["A", "E", "I", "O", "U"]

        for i in range(word_len):
            start_index += 1
            # check if 5 vowels in a row is still possible              
            if(word[i] == "A" and (word_len - start_index) < 4):
                return
            # append all vowels to a list
            elif(word[i] == "A" or word[i] == "E" or word[i] == "I" or 
            word[i] == "O" or word[i] == "U"):
                vowel_list.append(word[i])

        # check if first vowel in list is 'A'
        if(vowel_list[0] != "A"):
            return
        # remove duplicates from list
        for i in vowel_list:
            if i not in vowel_list_ol:
                vowel_list_ol.append(i)
        # check if all vowels in a row
        for i in range(len(vowels)):
            if(vowel_list_ol[i] != vowels[i]):
                return
        # if all checks passed print the word
        print(word, end="")
        
    for words in in_file:
        # check if all vowels are in words
        if("A" in words and "E" in words and "I" in words 
        and "O" in words and "U" in words):
            check_vowels(words)

if __name__ == '__main__':
    main()